### PR TITLE
ON HOLD (DO NOT MERGE) Add protobuf serializers

### PR DESCRIPTION
**Pull request by jmarin**
This PR adds custom serialization for persisted messages through `Akka Persistence`. The default serialization mechanism for `Akka` uses `Java Serialization`, which is not recommended for various reasons:

**1) Compatibility:** reading old events from the Journal can be problematic with a new version of the application. The binary representation of a class serialized with the default mechanism is not guaranteed to be compatible with an evolved version of that class.

**2) Security:** `Java Serialization` has been known to have security vulnerabilities, this [article](https://foxglovesecurity.com/2015/11/06/what-do-weblogic-websphere-jboss-jenkins-opennms-and-your-application-have-in-common-this-vulnerability/) does a good job at explaning the issues. The *TL;DR* version of this is: don't use `Java Serialization` and assume serializing / deserializing arrays of bytes with this mechanism is safe.

**2) Performance:** when actor messages are persisted with the default serializer based on `Java Persistence` a nice message like the following is printed out to the console:

```
[WARN] [10/05/2016 13:02:23.093] [test-system-akka.actor.default-dispatcher-2] 
[akka.serialization.Serialization(akka://test-system)] Using the default Java serializer for class 
[hmda.persistence.processing.HmdaFileValidator$QualityError] which is not recommended because 
of performance implications. Use another serializer or disable this warning using the setting 
'akka.actor.warn-about-java-serializer-usage'
```
To avoid these pitfalls, `Akka` provides an extensible mechanism to provide [Custom Serialization](http://doc.akka.io/docs/akka/current/scala/serialization.html). This PR adds custom serializers for persistent actors, changing the binary representation that is stored in the Journal. In order to allow easy schema evolution (i.e. changes in the `LAR` events from 2017 to 2018 model) a serialization mechanism that allows this in structured data is usually recommended. In this PR I have chosen [`Protocol Buffers`](https://developers.google.com/protocol-buffers/) for the following reasons:

* They are mature and widely used
* Already used internally by `Akka`
* Provide schema updates as built in capabilities (i.e. renaming field, adding fields, etc.) 
* Binary representation is compact and efficient, and plays well with backend journals (LevelDB, Cassandra)

Other options for solving the above mentioned issues are [`Avro`](https://avro.apache.org/), [`Thrift`](https://thrift.apache.org/) or plain `JSON` (though performance in this last case isn't on par, and schema evolution requires much more manual code).

A side benefit of using something like `Protocol Buffers` is that it decouples the persistence model from the domain model. For instance, in this PR the `Transmittal Sheet` and `Loan Application Register` have a flat structure in PBF, closer to the raw data that is transmitted when filing, while the domain model adds a richer and more fluent structure (there is the concept of a "loan", an "applicant", etc.). Converters between persistence and domain model are provided as implicit conversions in scope to the serializer. This will probably be changed in the future to use Event Adapters (see #610) that can also perform schema evolution (i.e. version 1 events to version 2 events). 

**Note 1:** This PR doesn't include a custom serializer for `HmdaFileValidator` messages yet, this [issue](https://github.com/cfpb/hmda-platform/issues/605) should be addressed first 

**Note 2:** When compiling the protobuf generated classes, some deprecation warnings are printed to the console. The `sbt` plugin used to generate `Scala` classes from `proto` files has addressed this in this [commit](https://github.com/trueaccord/ScalaPB/pull/157), which will go into their next release.

### CODE REVIEW

**Comment by schbetsy** [comment]
I haven't reviewed much of the code yet, but I wanted to say: Thanks for the thorough description! It's really helpful. :sparkles:

**Comment by kgudel** [comment]
I'd like to echo what @schbetsy has said! @jmarin what you've written up and the documentation provided is incredibly informative!

File: persistence/src/test/scala/hmda/persistence/processing/TestConfigOverride.scala [hunk 1]
```diff
@@ -10,6 +10,29 @@ object TestConfigOverride {
       | akka.persistence.journal.plugin = "inmemory-journal"
       | akka.persistence.query.journal.id = "inmemory-read-journal"
       | akka.persistence.snapshot-store.plugin = "inmemory-snapshot-store"
+      | akka.actor {
```
Comment (schbetsy) [inline persistence/src/test/scala/hmda/persistence/processing/TestConfigOverride.scala]: I'm interested in the difference between this Override test config and the normal test config (in `persistence/src/test/resources/application.conf`). 

What is the difference between the tests that have to use the override, and the ones that don't?

**Comment by schbetsy** [comment]
I'm seeing an error in the debug logs when I try to upload a file. I saw in the logs that the file successfully uploaded and parsed, then the log shows this: 
```
api [2016-10-07 12:28:33,215][DEBUG][hmda-akka.actor.default-dispatcher-22] Validation started for 1-2017-2
api [2016-10-07 12:28:33,227][DEBUG][hmda-akka.actor.default-dispatcher-5] request replay for persistenceId [HmdaFileParser-1-2017-2] from [0] to [9223372036854775807] limit [100]
api [2016-10-07 12:28:33,254][DEBUG][hmda-akka.actor.default-dispatcher-23] replay failed for persistenceId [HmdaFileParser-1-2017-2], due to [scala.Some cannot be cast to hmda.model.fi.lar.LoanApplicationRegister]
api [2016-10-07 12:28:33,263][DEBUG][hmda-akka.actor.default-dispatcher-23] request replay for persistenceId [HmdaFileParser-1-2017-2] from [0] to [9223372036854775807] limit [100]
api [2016-10-07 12:28:33,264][DEBUG][hmda-akka.actor.default-dispatcher-5] replay failed for persistenceId [HmdaFileParser-1-2017-2], due to [scala.Some cannot be cast to hmda.model.fi.lar.LoanApplicationRegister]
api [2016-10-07 12:29:24,200][DEBUG][hmda-akka.actor.default-dispatcher-5] Actor stopped at akka://hmda/user/supervisor/filings-1
api [2016-10-07 12:29:24,200][DEBUG][hmda-akka.actor.default-dispatcher-5] actor akka://hmda/user/supervisor/filings-1 terminated
api [2016-10-07 12:29:33,241][DEBUG][hmda-akka.actor.default-dispatcher-2] Actor stopped at akka://hmda/user/supervisor/submissions-1-2017
api [2016-10-07 12:29:33,241][DEBUG][hmda-akka.actor.default-dispatcher-2] actor akka://hmda/user/supervisor/submissions-1-2017 terminated
api [2016-10-07 12:29:35,821][DEBUG][hmda-akka.actor.default-dispatcher-2] Aborting tcp connection because of upstream failure: No elements passed in the last 1 minute.
api akka.stream.impl.Timers$IdleTimeoutBidi$$anon$7.onTimer(Timers.scala:160)
api akka.stream.stage.TimerGraphStageLogic.akka$stream$stage$TimerGraphStageLogic$$onInternalTimer(GraphStage.scala:1152)
api akka.stream.stage.TimerGraphStageLogic$$anonfun$akka$stream$stage$TimerGraphStageLogic$$getTimerAsyncCallback$1.apply(GraphStage.scala:1141)
...
```
The filing is still in status `validating`

**Comment by jmarin** [comment]
This definitely looks like a bug. I think I know what's going on, will look into it

**Comment by nickgrippin** [comment]
I'm getting a bunch of warnings when running the project a second time (without a `clean`):
```
[warn] /Users/grippinn/HMDA/hmda-platform/persistence/target/scala-2.11/src_managed/main/hmda/model/filing/FilingStatusMessage.scala:71: method valueOf in object FilingStatusMessage is deprecated: see corresponding Javadoc for more information.
[warn]   def toJavaValue(pbScalaSource: FilingStatusMessage): hmda.model.Filing.FilingStatusMessage = hmda.model.Filing.FilingStatusMessage.valueOf(pbScalaSource.value)
[warn]                                                                                                                                      ^
[warn] /Users/grippinn/HMDA/hmda-platform/persistence/target/scala-2.11/src_managed/main/hmda/model/institution/AgencyMessage.scala:91: method valueOf in object AgencyMessage is deprecated: see corresponding Javadoc for more information.
[warn]   def toJavaValue(pbScalaSource: AgencyMessage): hmda.model.Institution.AgencyMessage = hmda.model.Institution.AgencyMessage.valueOf(pbScalaSource.value)
[warn]                                                                                                                              ^
[warn] /Users/grippinn/HMDA/hmda-platform/persistence/target/scala-2.11/src_managed/main/hmda/model/institution/ExternalIdTypeMessage.scala:81: method valueOf in object ExternalIdTypeMessage is deprecated: see corresponding Javadoc for more information.
[warn]   def toJavaValue(pbScalaSource: ExternalIdTypeMessage): hmda.model.Institution.ExternalIdTypeMessage = hmda.model.Institution.ExternalIdTypeMessage.valueOf(pbScalaSource.value)
[warn]                                                                                                                                                      ^
[warn] /Users/grippinn/HMDA/hmda-platform/persistence/target/scala-2.11/src_managed/main/hmda/model/institution/InstitutionStatusMessage.scala:51: method valueOf in object InstitutionStatusMessage is deprecated: see corresponding Javadoc for more information.
[warn]   def toJavaValue(pbScalaSource: InstitutionStatusMessage): hmda.model.Institution.InstitutionStatusMessage = hmda.model.Institution.InstitutionStatusMessage.valueOf(pbScalaSource.value)
[warn]                                                                                                                                                               ^
[warn] /Users/grippinn/HMDA/hmda-platform/persistence/target/scala-2.11/src_managed/main/hmda/model/institution/InstitutionTypeMessage.scala:111: method valueOf in object InstitutionTypeMessage is deprecated: see corresponding Javadoc for more information.
[warn]   def toJavaValue(pbScalaSource: InstitutionTypeMessage): hmda.model.Institution.InstitutionTypeMessage = hmda.model.Institution.InstitutionTypeMessage.valueOf(pbScalaSource.value)
[warn]                                                                                                                                                         ^
[warn] 5 warnings found
```

**Comment by nickgrippin** [comment]
Error when uploading files (gets stuck on the `validating` state):
```
api [2016-10-13 16:32:24,252][INFO ][hmda-akka.actor.default-dispatcher-44] Message [akka.persistence.query.journal.leveldb.EventsByPersistenceIdPublisher$Continue$] 
from Actor[akka://hmda/user/StreamSupervisor-26/flow-26-1-currentEventsByPersistenceId-HmdaFileParser-1-2017-3-actorPublisherSource#216635216] 
to Actor[akka://hmda/user/StreamSupervisor-26/flow-26-1-currentEventsByPersistenceId-HmdaFileParser-1-2017-3-actorPublisherSource#216635216] was not delivered. 
[2] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
```

**Comment by jmarin** [comment]
I am putting this PR on hold. There are a few things that I don't love about the implementation, but more importantly, there are failures when running the whole system, and it is not obvious what is going on at first glance. So I think it's best to revisit this and take a fresh approach

**Comment by schbetsy** [comment]
I think that's fair. Thanks for being patient while we were figuring it all out.

**Pull request closed without merging**